### PR TITLE
feat:Spanner pg samples

### PR DESCRIPTION
**Pull request by Rishabh-V**
First draft of Spanner PostgreSQL dialect samples for feedback. The samples depend on v3.14.0 of C# Spanner client library which were created locally for sample creation and testing.

### CODE REVIEW

File: spanner/api/Spanner.Samples.Tests/UpdatePostgreSqlUsingBatchDmlAsyncTest.cs
Comment (Rishabh-V) [inline spanner/api/Spanner.Samples.Tests/UpdatePostgreSqlUsingBatchDmlAsyncTest.cs:1]: Please ignore this file from review. This is wrongly named and got checked-in in error.

File: spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs:28]: I see it is, thanks! Since we won't merge this until the library has been released and the feature is GA, maybe remove the setting of QuotaProject. We'll run testst when we are ready to merge and if they fail we can add it back in. Just don't want to put out there something that's not needed.

File: spanner/api/Spanner.Samples.Tests/CastDataTypesAsyncPostgreTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/CastDataTypesAsyncPostgreTest.cs:38]: Yep, avoid these tests on a constructed string.
Either return a struct from your tests or a value, like the amount of records fetched, etc.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Structure and overall looks good. Just a few comments.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:105]: This is used in one of the samples. I will share that in PR in second batch

File: spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs:51]: Right Amanda, this is indeed a requirement. This is based on the equivalent Java sample which I got to know after discussing with the team.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:436]: Initialize Database creates a PG database as well as couple of tables namely Singers and Albums. There is another sample which depicts the ordering of NULL (Nulls First, Nulls Last etc.) that creates a different version of Singers table and inserted data is then ordered in different ways. To keep the sample consistent and inline with other languages and to avoid the chances of duplicate data being inserted, I also created a CreateEmptyPostgreSqlDatabase method. I will also share this sample in next batch.

**Review by Rishabh-V — COMMENTED** [review]
Fixed review comments

File: spanner/api/Spanner.Samples/Spanner.Samples.csproj
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/Spanner.Samples.csproj:1]: This is still here. Revert the changes on this whole file.

File: spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs:51]: Acked. Thanks

File: spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs:40]: This is just a nit, but when you run this sample this text will be printed twice, once for creating the db and once for adding the tables. Consider distinguising the two as in "Waiting for the DB to be created" and "Waiting for the tables to be created."

Here and everywhere where there's something like this happening.

File: spanner/api/Spanner.Samples.Tests/CreateDatabaseAsyncPostgreTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/CreateDatabaseAsyncPostgreTest.cs:46]:
```suggestion
        Assert.Contains(databases, d => d.DatabaseName.DatabaseId == databaseId && d.DatabaseDialect == ....);
databaseId).DatabaseDialect);
```

File: spanner/api/Spanner.Samples/ConnectToDatabaseAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ConnectToDatabaseAsyncPostgre.cs:38]:
```suggestion
        var result = await command.ExecuteScalarAsync<string>();
        Console.WriteLine(result);
        return result;
```

File: spanner/api/Spanner.Samples/CastDataTypesAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/CastDataTypesAsyncPostgre.cs:68]: By the way, you have this on all your PRs, we usually leave a blank line at the end of the file.
Not a problem though, just fix it as you go along.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:105]: OK, so then should be included on the PR where the sample that uses it is. That way I can check the method and its usage, maybe suggest something else, etc. The same comment applies for the other methods that you are not using here but will use in another PR.

If it's not toooo much trouble, try to remove these unusued methods from this PR and add them on the subsequent ones. Else, no worries, but take it into account for next time.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:436]: OK, so then CreateEmpty... should be on the PR that uses it if posssible.

**Review by amanda-tarafa — COMMENTED** [review]
A few nits but otherwise looks good. Feel free to apply the changes here and create the PR containing the next batch (with relevant changes there as well acording to this PR's comments.)

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:105]: Sure Amanda. Added the test cases which uses this method.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:433]: Done.

File: spanner/api/Spanner.Samples/Spanner.Samples.csproj
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples/Spanner.Samples.csproj]: I reverted the entire file. However, still seeing it. I will seek your help to revert this tomorrow.

File: spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples/CreateDatabaseAsyncPostgre.cs]: Done.

File: spanner/api/Spanner.Samples/IdentifierCaseSensitivityAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/IdentifierCaseSensitivityAsyncPostgre.cs:123]: This is one of the main problems with having samples that do many things at once. We are barely testing this sample.

File: spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs:29]: Again, something that should be in docs and not on each of the language specific samples.

File: spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs:52]: What are the real types of these and why are you converting them to string?
I guess users would prefer to deal with something typed than with a string.

File: spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/GetInformationSchemaAsyncPostgre.cs:55]: Really don't use "null" or "NULL", that means nothing. Use `null`. If you absolutely need to provide a string value better use "unknown" or "undefined" as that's less prone to confusion. But in my view even that should only be used on a very explicit string representation like when writing to the Console or similar, never when holding the data.

File: spanner/api/Spanner.Samples/ExecutePartitionedDmlAsyncPostgre.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ExecutePartitionedDmlAsyncPostgre.cs:39]:
```suggestion
        Console.WriteLine($"At least {rowCount} row(s) deleted...");
```

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Apart from the specific change requests I have several PR wide comments.

1. There's a lot of information in comments inside the samples that should really be part of documentation sorrounding the sample instead. In that way, it doesn't have to be repeated across languages and it is maintained by TWs. Comments in samples should explain something particular of the language, i.e. something that is just needed for the .NET library for instance; and maybe small notes making reference to documentation, but not the whole explanation.
2. A lot of samples do many (related) things and we tend to prefer simpler samples. The current samples are harder to test and possibly harder for users to figure out (for instance the ORDER BY, the name Identifier samples, etc.)
3. Some samples create database and tables, some samples create tables and some samples asume everything they need is already created. We need consistency here and also sample simplification. If someone wants to see how to execute a command they probably don't need the whole create database => create table => execute command. If some samples are creating DBs to avoid flakyness, then those should be created in the corresponding tests but not on the samples themselves.
4. It's not clear to me that all samples and samples tests that are creating DBs are doing so because of flakyness. We are already blowing Spanner quota ocasionally, if we continut to create DBs without needing them these tests will never pass. Take a look at #1425. In addition consider creating a separate Fixture for PG.
5. See if you can reuse samples for the Spanner dialect, all of those that are the same, simply add the PG region tag on the existing one.

I know some of these issues are not because decisions you have made, but we should discus them nonetheless. Let's chat tomorrow.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (Rishabh-V) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:393]: `CreateEmptyPostgreSqlDatabaseAsync` and `ListTableNamesAsync` are now removed from fixture.

**Review by Rishabh-V — COMMENTED** [review]
Fixed review comments. The following changes are now done:
1. Removed comments that should be in documentation. Have kept only short summary now.
2. Refactored tests and samples, so that  new database is not created. To avoid data insertion conflicts, distinct data specific to the test scenario is inserted in tests as needed.

**Review by amanda-tarafa — APPROVED** [review]
Thanks Rishabh! These look good. LGTM

I'm marking this PR as `do not merge` just because we need production changes released first.

**Comment by amanda-tarafa** [comment]
@Rishabh-V  I think we might need to increase the overall timeout for Spanner samples. You can do that in the `runTests.ps1` file. Do include that change as part of this PR.

**Comment by amanda-tarafa** [comment]
> I'm marking this PR as do not merge just because we need production changes released first.

Is my assumption right that we still need to release some library changes or has everything been released already.

**Comment by Rishabh-V** [comment]
> Thanks Rishabh! These look good. LGTM
> 
> I'm marking this PR as `do not merge` just because we need production changes released first.

Thanks Amanda!

> @Rishabh-V I think we might need to increase the overall timeout for Spanner samples. You can do that in the `runTests.ps1` file. Do include that change as part of this PR.

Sure Amanda. Timeout is currently set to 7200 seconds.  Should I increase it to 9000 seconds?

> > I'm marking this PR as do not merge just because we need production changes released first.
> 
> Is my assumption right that we still need to release some library changes or has everything been released already.

Right Amanda. I believe PgNumeric is not released yet.

**Comment by amanda-tarafa** [comment]
Let's wait for the timeout until we can actually build the PR, as right now it is failing because it's missing PgNumeric as expected.

So, nothing to do for now. Thanks!

**Pull request merged**
